src/shared/trim-and-dedent.ts

```diff
@@ -1,24 +1,41 @@
+function trimLeadingEmptyLines(str: string): string {
+  return str.replace(/^\s*\n/g, '');
+}
+
+function trimTrailingEmptyLines(str: string): string {
+  return str.replace(/\n\s*$/g, '');
+}
+
 /**
- * Remove leading and trailing empty lines from a string.
+ * Remove specified indentation from each line of a string.
  */
-function trimEmptyLines(str: string): string {
-  return str.replace(/^\s*\n|\n\s*$/g, '');
+function dedentStr(str: string, indent: number) {
+  const indentRegexp = new RegExp(`^ {${indent}}`, 'gm');
+  return str.replace(indentRegexp, '');
 }
 
 /**
- * Remove common indentation from each line of a string.
+ * Remove common indentation from each line of every string, then interpolate
+ * params verbatim.
  */
-function dedent(str: string) {
-  // Match the smallest indentation
-  const match = str.match(/^[ \t]*(?=\S)/gm);
-  const indent = match && Math.min(...match.map(el => el.length));
-
-  if (indent) {
-    const regexp = new RegExp(`^ {${indent}}`, 'gm');
-    return str.replace(regexp, '');
+function dedent(strings: string[], ...params: any[]) {
+  // Match the smallest indentation among all strings
+  const indents = strings
+    .map(str => {
+      const match = str.match(/^[ \t]*(?=\S)/gm);
+      return match ? Math.min(...match.map(el => el.length)) : -1;
+    })
+    // Exclude lines where indentation could not be matched
+    .filter(num => num > -1);
```

Comment:
I wanted to use `null` or `undefined` as the "no-indentation-match` representation, but TS's definition of `filter` is to return the same type as the original array.

That means that, even if we know the filtering function is removing all instances of `null`, resulting on an array of only numbers, TS considers it is still an array of numbers or nulls, which then fails when passed to `Math.min`.

That's why I ended up using `-1` as the discriminator value, as it is still a number.

Reply:
Makes sense. Some other languages have a `filterMap` in their standard library which explicitly combines filtering and changing the type of the array. If JS had that it would make this kind of transform easier to figure out at the type level.